### PR TITLE
Introduce drop property tests

### PR DESCRIPTION
**Pull request by dbrajovic**
# Description

This PR imports Drop tests form the PBFT repo

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [ ] I have added sufficient documentation in code

### CODE REVIEW

File: core/drop_test.go [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+}
+
+func (c *cluster) runSequence(ctx context.Context, height uint64) {
+	for _, n := range c.nodes {
+		c.wg.Add(1)
+
+		go func(n *node) {
+			defer c.wg.Done()
+
+			if n.offline {
```
Comment (JekaMas) [inline core/drop_test.go:173]: All nodes are going to offline?

Comment (dbrajovic) [reply core/drop_test.go:173]: I don't think [so](https://github.com/0xPolygon/go-ibft/blob/b03b4031826dad456fbe2856ac59abefa3b1ac5d/core/drop_test.go#L51)

File: core/drop_test.go [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+				)
+			}
+		},
+	)
+
+	if err := cluster.progressToHeight(5*time.Second, 5); err != nil {
+		t.Fatal("cannot progress to height: err=", err)
+	}
+
+	cluster.stopN(int(cluster.maxFaulty()) + 1)
```
Comment (b00ris) [inline core/drop_test.go:51]: Can you use random nodes there?

File: core/drop_test.go [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+
+	assert.Error(t, cluster.progressToHeight(2*time.Second, 10))
```
Comment (b00ris) [inline core/drop_test.go:53]: It's worth to run them again at to check that consensus can continue

File: core/drop_test.go [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+		}
+
+		ok <- struct{}{}
+	}()
+
+	select {
+	case <-ok:
+		//	height reached
+	case <-ctx.Done():
+		return errors.New("timeout")
```
Comment (zivkovicmilos) [inline core/drop_test.go:205]: Nitick: you can probably extract this error out into a constant

File: core/drop_test.go [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+	return addresses
+}
+
+func (c *cluster) isProposer(
+	from []byte,
+	height,
+	round uint64,
+) bool {
+	if len(c.nodes) == 0 {
+		panic("no nodes")
```
Comment (zivkovicmilos) [inline core/drop_test.go:226]: Is this ever possible?

**Review by zivkovicmilos — APPROVED** [review]
Looks great 💯

Thank you for simplifying the code, the test looks clean 🙏

File: core/drop_test.go (hunk 2 above)
Comment (dbrajovic) [reply core/drop_test.go:51]: I didn't see any need for that in this particular case scenario. All the nodes are identical except for the [offline behavior](https://github.com/0xPolygon/go-ibft/blob/b03b4031826dad456fbe2856ac59abefa3b1ac5d/core/drop_test.go#L173), so having a random distribution does not affect the test flow.

**Comment by JekaMas** [comment]
Had to make my own fork as far as upstream is merging changes in months. Not mutation test framework uses modules and respect min MSI in config.
Also had to bump golang version because of failing govuln.

**Review by b00ris — APPROVED** [review]

**Comment by b00ris** [comment]
Looks good to me!

File: core/drop_test.go (hunk 4 above)
Comment (dbrajovic) [reply core/drop_test.go]: There's only one for the time being. I might add more depending on the use cases

File: core/drop_test.go (hunk 5 above)
Comment (dbrajovic) [reply core/drop_test.go]: [Removed]([e9a731a](https://github.com/0xPolygon/go-ibft/pull/31/commits/e9a731a265329995b4115c7e2fb05d14a28450d9))

File: core/drop_test.go [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+	}
+}
+
+func (c *cluster) maxFaulty() uint64 {
+	return (uint64(len(c.nodes)) - 1) / 3
+}
+
+func (c *cluster) stopN(num int) {
+	if num > len(c.nodes) {
+		panic(fmt.Sprintf("stop %d, but size is %d", num, len(c.nodes)))
```
[the remark this answers is not in the document]
Comment (dbrajovic) [reply core/drop_test.go]: [Removed](https://github.com/0xPolygon/go-ibft/pull/31/commits/e9a731a265329995b4115c7e2fb05d14a28450d9)

File: core/drop_test.go [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+	}
+
+	return nil
+}
+
+func (c *cluster) addresses() [][]byte {
+	addresses := make([][]byte, 0, len(c.nodes))
+	for _, node := range c.nodes {
+		addresses = append(addresses, node.address)
+	}
```
[the remark this answers is not in the document]
Comment (dbrajovic) [reply core/drop_test.go]: [Redone](https://github.com/0xPolygon/go-ibft/pull/31/commits/35096eb6c25740149da1481aec565e691f7efc2c)

File: core/drop_test.go [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+					},
+
+					&mockTransport{multicastFn: c.gossip},
+				)
+			}
+		},
+	)
+
+	if err := cluster.progressToHeight(5*time.Second, 5); err != nil {
+		t.Fatal("cannot progress to height: err=", err)
```
Comment (vcastellm) [inline core/drop_test.go:58]: `assert.NoError`?

File: core/drop_test.go [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
```
Comment (vcastellm) [inline core/drop_test.go:117]: `assert.NoError` with message?

File: core/drop_test.go [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+	return bytes.Equal(
+		proposal,
+		validProposal,
+	) && bytes.Equal(
+		proposalHash,
+		validProposalHash,
+	)
+}
+
+type node struct {
```
Comment (vcastellm) [inline core/drop_test.go:155]: Should this and cluster be moved to a separate file?

This drop test looks to me like e2e/integration tests, better to separate them as in https://github.com/0xPolygon/pbft-consensus/tree/main/e2e

File: core/drop_test.go (hunk 8 above)
Comment (dbrajovic) [reply core/drop_test.go]: [d60fe5c](https://github.com/0xPolygon/go-ibft/pull/31/commits/d60fe5cfd527d09c980d887fc1d3e1a5a76115a9)

File: core/drop_test.go (hunk 9 above)
Comment (dbrajovic) [reply core/drop_test.go]: [f96fe3f](https://github.com/0xPolygon/go-ibft/pull/31/commits/f96fe3ffeecd71eba9c8dd55f3b7025e408300f5)

File: core/drop_test.go (hunk 10 above)
Comment (dbrajovic) [reply core/drop_test.go]: [31fb07a](https://github.com/0xPolygon/go-ibft/pull/31/commits/31fb07a9a91e340a35a9e1545ef44d523aa93e29)

**Review by vcastellm — APPROVED** [review]
LGTM

**Pull request merged**